### PR TITLE
Initial version with String type instead of EventData and metrics HttpTrigger

### PR DESCRIPTION
**Pull request by eemhu**
* Changed EventData[] to String[] data type.
* Contains tests
* Contains access to prometheus metrics via metrics HttpTrigger
 -> Note: Access requires function key, can be provided via `?code=` query string or `x-functions-key` request header.
More details: https://learn.microsoft.com/en-us/azure/azure-functions/function-keys-how-to?tabs=azure-portal
* Uses Java 11.

### CODE REVIEW

File: local.settings.json
Comment (MoonBow-1) [inline local.settings.json:1]: Should this even be in the repository?

File: src/main/java/com/teragrep/aer_02/DefaultOutput.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.channels.UnresolvedAddressException;
+import java.util.concurrent.TimeoutException;
+
+import static com.codahale.metrics.MetricRegistry.name;
+
+// TODO unify, this is a copy from cfe_35 which is a copy from rlo_10 with FIXES
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/DefaultOutput.java:61]: New ticket about this, and then remove comment?

File: src/main/java/com/teragrep/aer_02/DefaultOutput.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        this.bytes = metricRegistry.counter(name(DefaultOutput.class, "<[" + name + "]>", "bytes"));
+        this.resends = metricRegistry.counter(name(DefaultOutput.class, "<[" + name + "]>", "resends"));
+        this.connects = metricRegistry.counter(name(DefaultOutput.class, "<[" + name + "]>", "connects"));
+        this.retriedConnects = metricRegistry.counter(name(DefaultOutput.class, "<[" + name + "]>", "retriedConnects"));
+        this.sendLatency = metricRegistry
+                .timer(name(DefaultOutput.class, "<[" + name + "]>", "sendLatency"), () -> new Timer(sendReservoir));
+        this.connectLatency = metricRegistry
+                .timer(name(DefaultOutput.class, "<[" + name + "]>", "connectLatency"), () -> new Timer(connectReservoir));
+
+        connect();
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/DefaultOutput.java:122]: Ctor is not code free, so can this be refactored to use **at least** the connect method outside the ctor?

File: src/main/java/com/teragrep/aer_02/EventDataConsumer.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        this.syslogConfig = new SyslogConfig(configSource);
+
+        this.jmxReporter = JmxReporter.forRegistry(metricRegistry).build();
+        this.slf4jReporter = Slf4jReporter
+                .forRegistry(metricRegistry)
+                .outputTo(LoggerFactory.getLogger(EventDataConsumer.class))
+                .convertRatesTo(TimeUnit.SECONDS)
+                .convertDurationsTo(TimeUnit.MILLISECONDS)
+                .build();
+        startMetrics();
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/EventDataConsumer.java:108]: Not code free

File: src/main/java/com/teragrep/aer_02/EventDataConsumer.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    private String getRealHostName() {
+        String hostname;
+        try {
+            hostname = InetAddress.getLocalHost().getHostName();
+        }
+        catch (UnknownHostException e) {
+            hostname = "localhost";
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/EventDataConsumer.java:117]: Warning or Error here?

File: src/main/java/com/teragrep/aer_02/EventDataConsumer.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    ) {
+
+        String partitionId = String.valueOf(partitionContext.get("PartitionId"));
+        metricRegistry.gauge(name(EventDataConsumer.class, "latency-seconds", partitionId), () -> new Gauge<Long>() {
+
+            @Override
+            public Long getValue() {
+                return Instant.now().getEpochSecond() - enqueuedTime.toInstant().getEpochSecond();
+            }
+        });
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/EventDataConsumer.java:146]: A lambda can be used here, but it makes the code a little less clear

```suggestion
        metricRegistry.gauge(name(EventDataConsumer.class, "latency-seconds", partitionId), () -> (Gauge<Long>) () ->
            Instant.now().getEpochSecond() - enqueuedTime.toInstant().getEpochSecond()
        );
```

File: src/main/java/com/teragrep/aer_02/EventDataConsumer.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+        String eventUuid = null; //TODO: Message id not available?
+
+        // FIXME proper handling of non-provided uuids
+        if (eventUuid == null) {
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/EventDataConsumer.java:151]: eventUuid is always null, check TODO comment

File: src/main/java/com/teragrep/aer_02/EventDataConsumer.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            eventUuid = "aer_01=" + UUID.randomUUID();
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/EventDataConsumer.java:152]: says AER_01, but should be AER_02

File: src/main/java/com/teragrep/aer_02/EventDataConsumer.java [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+                        "fully_qualified_namespace",
+                        String.valueOf(partitionContext.getOrDefault("FullyQualifiedNamespace", ""))
+                )
+                .addSDParam("eventhub_name", String.valueOf(partitionContext.getOrDefault("EventHubName", "")))
+                .addSDParam("partition_id", String.valueOf(partitionContext.getOrDefault("PartitionId", "")))
+                .addSDParam("consumer_group", String.valueOf(partitionContext.getOrDefault("ConsumerGroup", "")));
+
+        String partitionKey = String.valueOf(systemProps.getOrDefault("PartitionKey", ""));
+
+        // TODO: Correlation id not available?
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/EventDataConsumer.java:172]: New possible issue from TODO

File: src/main/java/com/teragrep/aer_02/EventDataConsumer.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+                .withSDElement(sdEvent)
+                //.withSDElement(sdCorId)
+                .withMsgId(String.valueOf(systemProps.getOrDefault("SequenceNumber", "0")))
+                .withMsg(eventData);
+
+        output.accept(syslogMessage.toRfc5424SyslogMessage().getBytes(StandardCharsets.UTF_8));
+    }
+
+    @Override
+    public void close() throws Exception {
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/EventDataConsumer.java:198]: Method never called?

File: src/main/java/com/teragrep/aer_02/config/AzureConfig.java
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/config/AzureConfig.java:1]: Changing all the fields to private revealed that the class is not accessed from outside, making it redundant.
Comments on this?

File: src/main/java/com/teragrep/aer_02/config/MetricsConfig.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+ * a licensee so wish it.
+ */
+package com.teragrep.aer_02.config;
+
+import com.teragrep.aer_02.config.source.Sourceable;
+
+public class MetricsConfig {
+
+    public final Sourceable configSource;
+    public final int prometheusPort;
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/config/MetricsConfig.java:53]: Public fields should be private

File: src/main/java/com/teragrep/aer_02/config/MetricsConfig.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+    public MetricsConfig(Sourceable configSource) {
+        this.configSource = configSource;
+        this.prometheusPort = Integer.parseInt(configSource.source("metrics.prometheusPort", "1234"));
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/config/MetricsConfig.java:57]: Not code free

File: src/main/java/com/teragrep/aer_02/config/RelpConfig.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+// copy from snw_01 with fixes
+public final class RelpConfig {
+
+    public final Sourceable configSource;
+    public final int connectionTimeout;
+    public final int readTimeout;
+    public final int writeTimeout;
+    public final int reconnectInterval;
+    public final int destinationPort;
+    public final String destinationAddress;
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/config/RelpConfig.java:59]: Not private

File: src/main/java/com/teragrep/aer_02/config/SyslogConfig.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+package com.teragrep.aer_02.config;
+
+import com.teragrep.aer_02.config.source.Sourceable;
+
+public final class SyslogConfig {
+
+    // TODO this is a copy from snw_01, with configSource
+    public final Sourceable configSource;
+    public final String hostname;
+    public final String appName;
```
Comment (MoonBow-1) [inline src/main/java/com/teragrep/aer_02/config/SyslogConfig.java:55]: Not private

File: src/test/java/com/teragrep/aer_02/ConfigTest.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+import com.teragrep.aer_02.config.SyslogConfig;
+import com.teragrep.aer_02.config.source.EnvironmentSource;
+import com.teragrep.aer_02.config.source.PropertySource;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+public class ConfigTest {
+
+    @Disabled(value = "fix maven")
```
Comment (MoonBow-1) [inline src/test/java/com/teragrep/aer_02/ConfigTest.java:59]: Disabled test?

File: src/main/java/com/teragrep/aer_02/config/AzureConfig.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * To the extent this program is licensed as part of the Commercial versions of
+ * Teragrep, the applicable Commercial License may apply to this file if you as
+ * a licensee so wish it.
+ */
+package com.teragrep.aer_02.config;
+
+import com.teragrep.aer_02.config.source.Sourceable;
+
+public final class AzureConfig {
```
Comment (kortemik) [inline src/main/java/com/teragrep/aer_02/config/AzureConfig.java:50]: is this needed here, propably not?

File: src/test/java/com/teragrep/aer_02/SyslogBridgeTest.java [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+    @AfterEach
+    void teardown() {
+        eventLoop.stop();
+        Assertions.assertDoesNotThrow(() -> eventLoopThread.join());
+        executorService.shutdown();
+        Assertions.assertDoesNotThrow(() -> server.close());
+        messages.clear();
+    }
+
+    @Test
```
Comment (kortemik) [inline src/test/java/com/teragrep/aer_02/SyslogBridgeTest.java:119]: test with ThrowingRelpConnectionFake too

**Review by kortemik — CHANGES_REQUESTED** [review]

**Review by kortemik — APPROVED** [review]

**Review by MoonBow-1 — CHANGES_REQUESTED** [review]

File: src/test/java/com/teragrep/aer_02/ConfigTest.java (hunk 14 above)
Comment (StrongestNumber9) [reply src/test/java/com/teragrep/aer_02/ConfigTest.java:59]: Also not a descriptive value, what is there to fix with maven?

**Review by MoonBow-1 — APPROVED** [review]

**Pull request merged**